Yubico.YubiKey/src/Yubico/YubiKey/Pipelines/CommandChainingTransform.cs

```diff
@@ -46,53 +47,77 @@ public ResponseApdu Invoke(CommandApdu command, Type commandType, Type responseT
+        {
+            ResponseApdu? responseApdu = null;
+            var sourceData = command.Data;
             while (!sourceData.IsEmpty)
             {
-                int chunkLength = Math.Min(MaxChunkSize, sourceData.Length);
-                var dataChunk = sourceData[..chunkLength];
-                sourceData = sourceData[chunkLength..];
-
-                var partialApdu = new CommandApdu
-                {
-                    Cla = (byte)(command.Cla | (sourceData.IsEmpty
-                        ? 0
-                        : 0x10)),
-                    Ins = command.Ins,
-                    P1 = command.P1,
-                    P2 = command.P2,
-                    Data = dataChunk
-                };
-
-                responseApdu = _pipeline.Invoke(partialApdu, commandType, responseType);
-
-                // Stop sending data when the YubiKey response is 0x6700 (when the chained data
-                // sent exceeds the max allowed length) and let caller handle the response
-                if (responseApdu.SW != SWConstants.WrongLength)
+                responseApdu = SendPartial(command, commandType, responseType, ref sourceData);
+                if (responseApdu.SW == SWConstants.Success)
                 {
                     continue;
                 }
 
-                _log.LogWarning("Sent data exceeds max allowed length by YubiKey. (SW: 0x{StatusWord})",
-                    responseApdu.SW.ToString("X4", CultureInfo.InvariantCulture));
-                
-                return responseApdu;
+                return FailedResponse(responseApdu);
             }
 
             return responseApdu!;
         }
 
-        public void Setup() => _pipeline.Setup();
+        private ResponseApdu SendPartial(
+            CommandApdu command,
+            Type commandType,
+            Type responseType,
+            ref ReadOnlyMemory<byte> sourceData)
+        {
+            int chunkLength = Math.Min(MaxChunkSize, sourceData.Length);
+            var dataChunk = sourceData[..chunkLength];
+            sourceData = sourceData[chunkLength..];
+
+            var partialApdu = new CommandApdu
+            {
+                Cla = (byte)(command.Cla | (sourceData.IsEmpty
+                    ? 0
+                    : 0x10)),
+                Ins = command.Ins,
+                P1 = command.P1,
+                P2 = command.P2,
+                Data = dataChunk
+            };
+
+            var responseApdu = _pipeline.Invoke(partialApdu, commandType, responseType);
+            return responseApdu;
+        }
+
+        private ResponseApdu FailedResponse(ResponseApdu responseApdu)
+        {
+            var currentCulture = CultureInfo.CurrentCulture;
+            string errorMessage = responseApdu.SW switch
+            {
+                SWConstants.WrongLength => string.Format(
+                    currentCulture, "Sent data exceeds max allowed length by YubiKey. (SW: 0x{0})",
```

Comment:
Technically, this might be false. There are other reasons we may get a WRONG_LENGTH 0x6700 SW from the YubiKey, so while the error *may* indicate that the amount of data sent exceeds the capacity, it's not guaranteed to be so.

Reply:
In the context of *any* APDU, generally speaking. It may be that the YubiKey specifically will always buffer the full request prior to parsing any of it, but that is not really a good assumption to make. Sure, that would allow you to infer that the issue is "too big" when there are still additional chunks to send, but that doesn't really help you when the final chuck is sent and you get that response (maybe the last fragment went over the limit, maybe it's just the response code from processing the command normally).